### ISSUE TITLE
Pathfinder

### ISSUE THREAD

**Issue by dirmeier** [pyro-ppl/numpyro#1485]
Hi all, are there any plans to implement pathfinder ([arXiv:2108.03782](https://arxiv.org/abs/2108.03782)) ? If so, I would be up for implementing it.

Cheers,
Simon

**Comment by fehiepsi** [comment]
Hi Simon, we don't have plan to implement that algorithm. Welcome to contribute!

**Comment by dirmeier** [comment]
Hi @xidulu , yeah I saw that some days after I wrote the original post and have been using it in the meantime. I still think it would be nice to have this in NumPyro, too, and I was planning on doing it over the christmas holidays.

**Comment by rlouf** [comment]
It is a shame to duplicate efforts between Blackjax and Numpyro when we could at the same time have a better PPL _and_ more samplers 😁

**Comment by fehiepsi** [comment]
@dirmeier Please feel free to work on this if you find that it is easier to use when having a numpyro implementation. Or defer to blackjax if you find that no need to duplicate efforts. Thanks!

@xidulu `jaxopt` has [LBFGS](https://jaxopt.github.io/stable/_autosummary/jaxopt.LBFGS.html) implemented. We're trying to do some benchmarks on the top of it to see if it's fast (upon batching) and stable. I'm not sure if it is flexible enough for pathfinder.

Good to know that blackjax has this sampler. It is great for the community.

**Comment by fehiepsi** [comment]
@xidulu Looking at your script, it seems that the [MCMCKernel](https://num.pyro.ai/en/stable/mcmc.html#numpyro.infer.mcmc.MCMCKernel) api can serve as a wrapper for it - there we can
+ converting a model to potential function
+ perform an update step and scan the samples
I guess this pattern works for many blackjax samplers. Somehow this is similar to how we wrap [TFPKernel](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/contrib/tfp/mcmc.py) to use TFP samplers for numpyro models.

This way we can take some benefits of [MCMC](https://num.pyro.ai/en/stable/mcmc.html#numpyro.infer.mcmc.MCMC) class which allows us to get unconstrained samples, deterministic samples, diagnostics, vectorized/parallel chains, progress bar, thinning,... [ArviZ](https://python.arviz.org/en/stable/) is also compatible with MCMC class so we can do further exploratory analysis.

**Comment by juanitorduz** [comment]
Hi! I'm trying to use blackjax's pathfinder method with numpyro with above. This gets me the samples in the transformed space; how can I get them to the original space? Thanks

I have tried 

```python
samples_transformed = jax.vmap( transform_func(*args, **kwargs))(samples)
```

But got non-sense values ...

**Comment by fehiepsi** [comment]
So you're using `dynamics_args=True`, which is different. Then `postprocess_fn` should do such a transform. Did you check your blackjax results? You may also want to check if you get correct results with manual transforms. Also check for what you provide to blackjax, like whether `logdensity_fn` has correct sign.

**Comment by juanitorduz** [comment]
> So you're using `dynamics_args=True`, which is different. Then `postprocess_fn` should do such a transform. Did you check your blackjax results? You may also want to check if you get correct results with manual transforms. Also check for what you provide to blackjax, like whether `logdensity_fn` has correct sign.

OMG 🤦  , actually

```python
def logdensity_fn(position):
    func = potential_fn(x, y)
    return func(position)
```
had the right sign and now is working as expected:

![image](https://github.com/pyro-ppl/numpyro/assets/22996444/65753fb7-a07a-4a19-9be4-4ff539d2c239)

and 

```python
posterior_samples_transformed

>> {'a': Array([0.9645677 , 0.88383734, 0.9891541 , ..., 0.98004097, 0.9882706 ,
        1.0226959 ], dtype=float32),
 'b': Array([1.9266534, 1.9878191, 1.8892442, ..., 1.8811033, 1.9484416,
        1.9543549], dtype=float32),
 'sigma': Array([0.49401087, 0.49979746, 0.46184635, ..., 0.44290245, 0.51779646,
        0.49234945], dtype=float32)}
```

I hope this can help other users :) Thanks!

Is there is still interest in having a wrapper around [MCMCKernel](https://num.pyro.ai/en/stable/mcmc.html#numpyro.infer.mcmc.MCMCKernel) ?

**Comment by juanitorduz** [comment]
> @xidulu Looking at your script, it seems that the [MCMCKernel](https://num.pyro.ai/en/stable/mcmc.html#numpyro.infer.mcmc.MCMCKernel) api can serve as a wrapper for it - there we can
> 
> * converting a model to potential function
> * perform an update step and scan the samples
>   I guess this pattern works for many blackjax samplers. Somehow this is similar to how we wrap [TFPKernel](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/contrib/tfp/mcmc.py) to use TFP samplers for numpyro models.
> 
> This way, we can take some benefits of [MCMC](https://num.pyro.ai/en/stable/mcmc.html#numpyro.infer.mcmc.MCMC) class, which allows us to get unconstrained samples, deterministic samples, diagnostics, vectorized/parallel chains, progress bar, thinning,... [ArviZ](https://python.arviz.org/en/stable/) is also compatible with MCMC class so we can do further exploratory analysis.

@fehiepsi  Is this still open? Meaning, wrapping this in the MCMC class? Or is it enough to have an example in the documentation?

**Comment by fehiepsi** [comment]
Hi @juanitorduz I think it be great to have a tutorial / example showing how to use numpyro with other libraries via initialize_model. If you think that the MCMC class is helpful, you can also show in the example how to define a MCMCKernel for the inference algorithm.

The issue https://github.com/pyro-ppl/numpyro/issues/1734 is also related.

**Comment by juanitorduz** [comment]
Ok! I can work on the example! I actually wrote something for myself based on this issue https://juanitorduz.github.io/numpyro_pathfinder/ is this example too easy? Shall I do something more complex? I did an easy one not to get lost in the model but rather in the Pathfinder part. But maybe NumPyro’s audience wants a less “trivial” example. I'm easy :)

**Comment by fehiepsi** [comment]
I would prefer simple examples to complex ones. :)

**Comment by juanitorduz** [comment]
> I would prefer simple examples to complex ones. :)

Same here 😀 !

**Comment by fehiepsi** [comment]
please check out https://num.pyro.ai/en/latest/tutorials/other_samplers.html for how to use numpyro with other apis

**Comment by seongwoohan** [comment]
Hi @juanitorduz , I've been using your pathfinder implementation. This is amazing! I think this implementation is singple-path pathfinder run. Do you think it's possible to run multiple-path pathfinder following algorithm2 in the pathfinder paper (https://jmlr.org/papers/volume23/21-0889/21-0889.pdf)? I think in that case, we can run your algorithms multiple times with different initializations in parallel and then apply Arviz's Pareto smoothed importance sampling.
